File: packages/sdk-typescript/src/gateway.ts
import { AgentVoirError } from "./errors.js";
import type { ChatCompletionRequest, JsonObject, UsageEventRequest } from "./types.js";

const SDK_VERSION = "0.1.0";

export interface GatewayClientOptions {
  baseUrl: string;
  apiKey: string;
  agentId: string;
  agentVersion?: string;
  tenantId?: string;
  userId?: string;
  fetchImpl?: typeof fetch;
}

export class GatewayClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly agentId: string;
  private readonly agentVersion: string;
  private readonly tenantId: string;
  private readonly userId?: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: GatewayClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.apiKey = options.apiKey;
    this.agentId = options.agentId;
    this.agentVersion = options.agentVersion ?? "0.1.0";
    this.tenantId = options.tenantId ?? "default";
    this.userId = options.userId;
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async chatCompletions(body: ChatCompletionRequest): Promise<JsonObject> {
    const response = await this.fetchImpl(`${this.baseUrl}/v1/chat/completions`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(body),
    });
    if (!response.ok) {
      throw new AgentVoirError(await response.text(), response.status);
    }
    return (await response.json()) as JsonObject;
  }

  async listModels(): Promise<JsonObject> {
    const response = await this.fetchImpl(`${this.baseUrl}/v1/models`, {
      headers: this.headers(false),
    });
    if (!response.ok) {
      throw new AgentVoirError(await response.text(), response.status);
    }
    return (await response.json()) as JsonObject;
  }

  private headers(includeContentType = true): Record<string, string> {
    const headers: Record<string, string> = {
      authorization: `Bearer ${this.apiKey}`,
      "x-agent-id": this.agentId,
      "x-agent-version": this.agentVersion,
      "x-tenant-id": this.tenantId,
      "user-agent": `agentvoir-typescript-sdk/${SDK_VERSION}`,
    };
    if (includeContentType) {
      headers["content-type"] = "application/json";
    }
    if (this.userId) {
      headers["x-user-id"] = this.userId;
    }
    return headers;
  }
}

export interface UsageClientOptions {
  baseUrl: string;
  fetchImpl?: typeof fetch;
}

export class UsageClient {
  private readonly baseUrl: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: UsageClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, "");
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async ingestEvent(body: UsageEventRequest): Promise<JsonObject> {
    const response = await this.fetchImpl(`${this.baseUrl}/v1/usage-events`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });
    if (!response.ok) {
      throw new AgentVoirError(await response.text(), response.status);
    }
    return (await response.json()) as JsonObject;
  }

  async listEvents(options: {
    agentId?: string;
    tenantId?: string;
    limit?: number;
  } = {}): Promise<JsonObject[]> {
    const url = new URL(`${this.baseUrl}/v1/usage-events`);
    if (options.agentId) url.searchParams.set("agent_id", options.agentId);
    if (options.tenantId) url.searchParams.set("tenant_id", options.tenantId);
    url.searchParams.set("limit", String(options.limit ?? 100));

    const response = await this.fetchImpl(url);
    if (!response.ok) {
      throw new AgentVoirError(await response.text(), response.status);
    }
    return (await response.json()) as JsonObject[];
  }
}

File: packages/sdk-python/src/agentvoir/gateway.py
from __future__ import annotations

from typing import Any

import httpx

from .errors import AgentVoirError
from .types import ChatCompletionRequest, JsonObject, UsageEventRequest

SDK_VERSION = "0.1.0"


class GatewayClient:
    """OpenAI-compatible client for the AgentVoir LLM gateway."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        *,
        agent_id: str,
        agent_version: str = "0.1.0",
        tenant_id: str = "default",
        user_id: str = "",
        timeout: float = 60.0,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.agent_id = agent_id
        self.agent_version = agent_version
        self.tenant_id = tenant_id
        self.user_id = user_id
        self.timeout = timeout

    def _headers(self) -> dict[str, str]:
        headers = {
            "authorization": f"Bearer {self.api_key}",
            "content-type": "application/json",
            "x-agent-id": self.agent_id,
            "x-agent-version": self.agent_version,
            "x-tenant-id": self.tenant_id,
            "user-agent": f"agentvoir-python-sdk/{SDK_VERSION}",
        }
        if self.user_id:
            headers["x-user-id"] = self.user_id
        return headers

    def chat_completions(self, request: ChatCompletionRequest) -> JsonObject:
        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                f"{self.base_url}/v1/chat/completions",
                headers=self._headers(),
                json=request.model_dump(exclude_none=True),
            )
        if response.status_code >= 400:
            raise AgentVoirError(response.text.strip(), response.status_code)
        return response.json()

    def list_models(self) -> JsonObject:
        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(
                f"{self.base_url}/v1/models",
                headers=self._headers(),
            )
        if response.status_code >= 400:
            raise AgentVoirError(response.text.strip(), response.status_code)
        return response.json()


class UsageClient:
    """Client for the AgentVoir token-accounting / usage ingestion API."""

    def __init__(self, base_url: str, timeout: float = 30.0) -> None:
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    def ingest_event(self, request: UsageEventRequest) -> JsonObject:
        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                f"{self.base_url}/v1/usage-events",
                headers={"content-type": "application/json"},
                json=request.model_dump(exclude_none=True),
            )
        if response.status_code >= 400:
            raise AgentVoirError(response.text.strip(), response.status_code)
        return response.json()

    def list_events(
        self,
        *,
        agent_id: str | None = None,
        tenant_id: str | None = None,
        limit: int = 100,
    ) -> list[JsonObject]:
        params: dict[str, Any] = {"limit": limit}
        if agent_id:
            params["agent_id"] = agent_id
        if tenant_id:
            params["tenant_id"] = tenant_id
        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(
                f"{self.base_url}/v1/usage-events",
                params=params,
            )
        if response.status_code >= 400:
            raise AgentVoirError(response.text.strip(), response.status_code)
        return response.json()
